### PR TITLE
graceful shutdown

### PR DESCRIPTION
**Pull request by hshomroni**
fix delayed destruction of DBChcheckppint manager due to sleeping monitor thread

### CODE REVIEW

**Comment by glevkovich** [comment]
@hshomroni , general comment about the commit message: there are some typos, and I would format it differently:
`DbCheckpointManager:fix forced kill of concord (sleeping monitor thread)`

I keep it short to go on the 80 chars lim
Keeping header short and clear, is always more important than details.. details should be given in body.
After that I would add a commit body, of 1-2-3 lines explaining why it was killed forcibly.

File: bftengine/include/bftengine/DbCheckpointManager.hpp [hunk 1]
```diff
@@ -196,6 +197,8 @@ class DbCheckpointManager {
   mutable std::mutex lock_;
   mutable std::mutex lockDbCheckPtFuture_;
   std::mutex lockLastDbCheckpointDesc_;
+  std::condition_variable shutdownCond_;  // used to wake up monitor thread periodically or when destructing an obj
+  std::mutex lockShutdown_;          // used for the above cond var. doesn't really coordinate anything between threads.
```
Comment (glevkovich) [inline bftengine/include/bftengine/DbCheckpointManager.hpp]: I think that if you ended with cond, you should end with lock. Also, the comment of this line is completely redundent in my opinion for anyone with basic c++ knowledge.
`std::mutex shutdownLock_;`

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 2]
```diff
@@ -135,8 +135,12 @@ void DbCheckpointManager::init() {
 
   monitorThread_ = std::thread([this]() {
     while (maxNumOfCheckpoints_ && !stopped_) {
+      std::unique_lock<std::mutex> lk(lockShutdown_);
       checkAndRemove();
-      std::this_thread::sleep_for(std::chrono::seconds{ReplicaConfig::instance().dbCheckpointMonitorIntervalSeconds});
+      shutdownCond_.wait_for(lk,
+                             std::chrono::seconds{ReplicaConfig::instance().dbCheckpointMonitorIntervalSeconds},
+                             [this] { return stopped_ == true; });
+      lk.unlock();
```
Comment (glevkovich) [inline bftengine/src/bftengine/DbCheckpointManager.cpp]: redundent.
unique_lock automatically unlock the mutex it holds on distructor.
You should remove this one.
see: https://en.cppreference.com/w/cpp/thread/unique_lock/~unique_lock

**Review by glevkovich — APPROVED** [review]
@hshomroni  Thanks for fixing this one! , approved. Please see my 2 comments. If you have questions or do not aggree with my comments - lets discuss on zoom.

File: bftengine/include/bftengine/DbCheckpointManager.hpp (hunk 1 above)
Comment (hshomroni) [reply bftengine/include/bftengine/DbCheckpointManager.hpp]: done

File: bftengine/src/bftengine/DbCheckpointManager.cpp (hunk 2 above)
Comment (hshomroni) [reply bftengine/src/bftengine/DbCheckpointManager.cpp]: done

**Pull request merged**